File: frontend/src/api/adminRbac.ts
import api from './index'

const adminOpts = { skipErrorToast: true } as const

export type DepartmentItem = {
  id: number
  name: string
  parent_id: number | null
  sort_order: number
  created_at: string
  is_builtin: boolean
}

export type GroupItem = {
  id: number
  name: string
  description: string | null
  created_at: string
}

export type EnterpriseRoleItem = {
  id: number
  slug: string
  name: string
  description: string | null
  is_builtin: boolean
  is_active: boolean
  created_at: string
}

export type EnterpriseRoleDeleteSummary = {
  deleted_user_role_assignments: number
  deleted_acl_rows: number
  message: string
}

export type UserOrgGroupItem = {
  id: number
  name: string
}

export type AdminUserOrg = {
  user_id: number
  primary_department_id: number
  primary_department_name: string
  groups: UserOrgGroupItem[]
}

export type WorkspaceMemberRoles = {
  user_id: number
  role_ids: number[]
  role_slugs: string[]
}

export type FolderAclEntryItem = {
  id: number
  folder_id: number | null
  subject_type: 'user' | 'role' | 'group' | 'department'
  subject_id: number
  permission: 'list' | 'read' | 'write' | 'manage'
  created_at: string
  updated_at: string
}

export type FolderAclEntryInput = {
  folder_id: number | null
  subject_type: 'user' | 'role' | 'group' | 'department'
  subject_id: number
  permission: 'list' | 'read' | 'write' | 'manage'
}

export type FolderAclPutSummary = {
  upserted: number
  updated: number
}

export function listAdminDepartments() {
  return api.get<DepartmentItem[]>('/admin/departments', adminOpts)
}

export function createAdminDepartment(body: { name: string; parent_id: number; sort_order?: number }) {
  return api.post<DepartmentItem>('/admin/departments', body, adminOpts)
}

export function updateAdminDepartment(
  id: number,
  body: { name?: string; parent_id?: number | null; sort_order?: number },
) {
  return api.put<DepartmentItem>(`/admin/departments/${id}`, body, adminOpts)
}

export function deleteAdminDepartment(id: number) {
  return api.delete(`/admin/departments/${id}`, adminOpts)
}

export function listAdminGroups() {
  return api.get<GroupItem[]>('/admin/groups', adminOpts)
}

export function createAdminGroup(body: { name: string; description?: string | null }) {
  return api.post<GroupItem>('/admin/groups', body, adminOpts)
}

export function updateAdminGroup(id: number, body: { name?: string; description?: string | null }) {
  return api.put<GroupItem>(`/admin/groups/${id}`, body, adminOpts)
}

export function deleteAdminGroup(id: number) {
  return api.delete(`/admin/groups/${id}`, adminOpts)
}

export function listEnterpriseRoles() {
  return api.get<EnterpriseRoleItem[]>('/admin/enterprise-roles', adminOpts)
}

export function createEnterpriseRole(body: { slug: string; name: string; description?: string | null }) {
  return api.post<EnterpriseRoleItem>('/admin/enterprise-roles', body, adminOpts)
}

export function updateEnterpriseRole(
  id: number,
  body: { name?: string; description?: string | null; is_active?: boolean },
) {
  return api.put<EnterpriseRoleItem>(`/admin/enterprise-roles/${id}`, body, adminOpts)
}

export function deleteEnterpriseRole(id: number) {
  return api.delete<EnterpriseRoleDeleteSummary>(`/admin/enterprise-roles/${id}`, adminOpts)
}

export function getAdminUserOrg(userId: number) {
  return api.get<AdminUserOrg>(`/admin/users/${userId}/org`, adminOpts)
}

export function putAdminUserOrg(userId: number, body: { primary_department_id: number; group_ids: number[] }) {
  return api.put<AdminUserOrg>(`/admin/users/${userId}/org`, body, adminOpts)
}

export function getWorkspaceMemberRoles(workspaceId: number, userId: number) {
  return api.get<WorkspaceMemberRoles>(
    `/admin/workspaces/${workspaceId}/members/${userId}/roles`,
    adminOpts,
  )
}

export function putWorkspaceMemberRoles(workspaceId: number, userId: number, roleIds: number[]) {
  return api.put<WorkspaceMemberRoles>(
    `/admin/workspaces/${workspaceId}/members/${userId}/roles`,
    { role_ids: roleIds },
    adminOpts,
  )
}

export function listWorkspaceFolderAcl(workspaceId: number) {
  return api.get<FolderAclEntryItem[]>(`/admin/workspaces/${workspaceId}/folder-acl`, adminOpts)
}

export function putWorkspaceFolderAcl(workspaceId: number, entries: FolderAclEntryInput[]) {
  return api.put<FolderAclPutSummary>(
    `/admin/workspaces/${workspaceId}/folder-acl`,
    { entries },
    adminOpts,
  )
}

File: backend/schemas/admin_rbac.py
# Copyright (c) 2026 徐泽宇
"""059 P2 管理端 RBAC API schemas。"""

from __future__ import annotations

from pydantic import BaseModel, Field


class DepartmentResponse(BaseModel):
    id: int
    name: str
    parent_id: int | None
    sort_order: int
    created_at: str
    is_builtin: bool = False


class DepartmentCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    parent_id: int
    sort_order: int = 0


class DepartmentUpdateRequest(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    parent_id: int | None = None
    sort_order: int | None = None


class GroupResponse(BaseModel):
    id: int
    name: str
    description: str | None
    created_at: str


class GroupCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)


class GroupUpdateRequest(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)


class EnterpriseRoleResponse(BaseModel):
    id: int
    slug: str
    name: str
    description: str | None
    is_builtin: bool
    is_active: bool
    created_at: str


class EnterpriseRoleCreateRequest(BaseModel):
    slug: str = Field(..., min_length=1, max_length=64, pattern=r"^[a-z][a-z0-9_]*$")
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)


class EnterpriseRoleUpdateRequest(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    is_active: bool | None = None


class EnterpriseRoleDeleteResponse(BaseModel):
    deleted_user_role_assignments: int
    deleted_acl_rows: int
    message: str


class UserOrgGroupItem(BaseModel):
    id: int
    name: str


class AdminUserOrgResponse(BaseModel):
    user_id: int
    primary_department_id: int
    primary_department_name: str
    groups: list[UserOrgGroupItem]


class AdminUserOrgUpdateRequest(BaseModel):
    primary_department_id: int
    group_ids: list[int] = Field(default_factory=list)


class WorkspaceMemberRolesUpdateRequest(BaseModel):
    role_ids: list[int] = Field(default_factory=list)


class WorkspaceMemberRolesResponse(BaseModel):
    user_id: int
    role_ids: list[int]
    role_slugs: list[str]
    rollback_warning: bool = False


class FolderAclEntryInput(BaseModel):
    folder_id: int | None = None
    subject_type: str = Field(..., pattern=r"^(user|role|group|department)$")
    subject_id: int
    permission: str = Field(..., pattern=r"^(list|read|write|manage)$")


class FolderAclSubjectEntryInput(BaseModel):
    subject_type: str = Field(..., pattern=r"^(user|role|group|department)$")
    subject_id: int
    permission: str = Field(..., pattern=r"^(list|read|write|manage)$")


class FolderAclEntryResponse(BaseModel):
    id: int
    folder_id: int | None
    subject_type: str
    subject_id: int
    permission: str
    created_at: str
    updated_at: str


class FolderAclBulkPutRequest(BaseModel):
    entries: list[FolderAclEntryInput] = Field(default_factory=list)


class FolderAclFolderPutRequest(BaseModel):
    entries: list[FolderAclSubjectEntryInput] = Field(default_factory=list)


class FolderAclPutSummaryResponse(BaseModel):
    upserted: int
    updated: int
    rollback_warning: bool = False
